pkg/uploader/util.go

```diff
@@ -0,0 +1,112 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package uploader
+
+import (
+	"strings"
+
+	corev1api "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+)
+
+const (
+	// PVCNameAnnotation is the key for the annotation added to
+	// pod volume backups when they're for a PVC.
+	PVCNameAnnotation = "velero.io/pvc-name"
+
+	// Deprecated.
+	//
+	// TODO(2.0): remove
+	podAnnotationPrefix = "snapshot.velero.io/"
+)
+
+// GetVolumeBackupsForPod returns a map, of volume name -> snapshot id,
+// of the PodVolumeBackups that exist for the provided pod.
+func GetVolumeBackupsForPod(podVolumeBackups []*velerov1api.PodVolumeBackup, pod *corev1api.Pod, sourcePodNs string) map[string]string {
+	volumes := make(map[string]string)
+
+	for _, pvb := range podVolumeBackups {
+		if !isPVBMatchPod(pvb, pod.GetName(), sourcePodNs) {
+			continue
+		}
+
+		// skip PVBs without a snapshot ID since there's nothing
+		// to restore (they could be failed, or for empty volumes).
+		if pvb.Status.SnapshotID == "" {
+			continue
+		}
+
+		// If the volume came from a projected or DownwardAPI source, skip its restore.
+		// This allows backups affected by https://github.com/vmware-tanzu/velero/issues/3863
+		// or https://github.com/vmware-tanzu/velero/issues/4053 to be restored successfully.
+		if volumeHasNonRestorableSource(pvb.Spec.Volume, pod.Spec.Volumes) {
+			continue
+		}
+
+		volumes[pvb.Spec.Volume] = pvb.Status.SnapshotID
+	}
+
+	if len(volumes) > 0 {
+		return volumes
+	}
+
+	return getPodSnapshotAnnotations(pod)
+}
+
+func isPVBMatchPod(pvb *velerov1api.PodVolumeBackup, podName string, namespace string) bool {
+	return podName == pvb.Spec.Pod.Name && namespace == pvb.Spec.Pod.Namespace
+}
+
+// volumeHasNonRestorableSource checks if the given volume exists in the list of podVolumes
+// and returns true if the volume's source is not restorable. This is true for volumes with
+// a Projected or DownwardAPI source.
+func volumeHasNonRestorableSource(volumeName string, podVolumes []corev1api.Volume) bool {
+	var volume corev1api.Volume
+	for _, v := range podVolumes {
+		if v.Name == volumeName {
+			volume = v
+			break
+		}
+	}
+	return volume.Projected != nil || volume.DownwardAPI != nil
+}
+
+// getPodSnapshotAnnotations returns a map, of volume name -> snapshot id,
+// of all restic snapshots for this pod.
+// TODO(2.0) to remove
+// Deprecated: we will stop using pod annotations to record restic snapshot IDs after they're taken,
+// therefore we won't need to check if these annotations exist.
+func getPodSnapshotAnnotations(obj metav1.Object) map[string]string {
```

Comment:
Change the comments here? Replace "restic"

Reply:
It's hard to replace all "restic" in the code/comments until we reach an agreement on what should be changed to. Here  is already a open [issue](https://github.com/vmware-tanzu/velero/issues/5094) to track the task, let's do the replacement when handling the issue

Reply:
Ok, thanks.